File: bench/sources/bench.cpp
#include <stdio.h>
#include <cstring>
#include <vector>
#include <string>
#include <cmath>

long long fib(int n) {
    if (n <= 1) return n;
    return fib(n - 1) + fib(n - 2);
}

int main(int argc, char *argv[]) {
    if (argc < 2) { fprintf(stderr, "Usage: bench <test>\n"); return 1; }
    const char *test = argv[1];

    if (strcmp(test, "fib") == 0) {
        printf("%lld\n", fib(35));
    } else if (strcmp(test, "loop") == 0) {
        long long total = 0;
        for (long long i = 0; i < 100000000LL; i++) total += i;
        printf("%lld\n", total);
    } else if (strcmp(test, "array") == 0) {
        std::vector<long long> arr;
        arr.reserve(1000000);
        for (int i = 0; i < 1000000; i++) arr.push_back(i);
        printf("%d\n", (int)arr.size());
    } else if (strcmp(test, "nested") == 0) {
        long long total = 0;
        for (int i = 0; i < 500; i++)
            for (int j = 0; j < 500; j++)
                total += (long long)i * j;
        printf("%lld\n", total);
    } else if (strcmp(test, "string") == 0) {
        std::string s;
        s.reserve(100001);
        for (int i = 0; i < 100000; i++) s += 'a';
        printf("%d\n", (int)s.size());
    } else if (strcmp(test, "math") == 0) {
        double total = 0.0;
        for (int i = 0; i < 1000000; i++) total += std::sqrt((double)(i + 1));
        printf("%.0f\n", total);
    }
    return 0;
}

File: bench/sources/bench.rs
use std::io::{self, Write};

fn fib(n: i64) -> i64 {
    if n <= 1 { return n; }
    fib(n - 1) + fib(n - 2)
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    if args.len() < 2 { eprintln!("Usage: bench <test>"); std::process::exit(1); }
    let test = &args[1];

    if test == "fib" {
        println!("{}", fib(35));
    } else if test == "loop" {
        let mut total: i64 = 0;
        for i in 0..100_000_000i64 { total += i; }
        println!("{}", total);
    } else if test == "array" {
        let mut arr: Vec<i64> = Vec::with_capacity(1_000_000);
        for i in 0..1_000_000 { arr.push(i); }
        println!("{}", arr.len());
    } else if test == "nested" {
        let mut total: i64 = 0;
        for i in 0..500i64 { for j in 0..500i64 { total += i * j; } }
        println!("{}", total);
    } else if test == "string" {
        let mut s = String::with_capacity(100_001);
        for _ in 0..100_000 { s.push('a'); }
        println!("{}", s.len());
    } else if test == "math" {
        let mut total: f64 = 0.0;
        for i in 0..1_000_000 { total += ((i + 1) as f64).sqrt(); }
        println!("{:.0}", total);
    }
}
